src/Storages/StorageS3.h

```diff
@@ -302,8 +318,9 @@ class StorageS3 : public IStorage, WithContext
 private:
     friend class StorageS3Cluster;
     friend class TableFunctionS3Cluster;
-    friend class StorageHudi;
-    friend class StorageDeltaLake;
+    friend class IStorageDataLake<StorageHudiName, HudiMetadataParser<StorageS3::Configuration, S3DataLakeMetadataReadHelper>>;
```

Comment:
Not needed anymore. The method has been moved to public.

Reply:
btw, I now had a quick idea/question : why don't we inherit IStorageDeltaLake from template Storage (S3Storage), would it be more clean and readable? I am not sure at the moment whether it will be better, but at least then there will be not need in moving methods to `public`, they could be `protected`

Reply:
And also it will allow to get rid of template configuration structure, as we just will be able to use `Storage::Configuration` instead.

Reply:
Maybe it's not a good idea. For example, the IStorageDataLake does not support `write` and `truncate`, also does not have virtual column, so we also need to override these related methods.

Reply:
> For example, the IStorageDataLake does not support write and truncate, also does not have virtual column, so we also need to override these related methods.

override and throw exception "Not supported"? It seems better than having a lot of templates for everything: for Storage, for MetadataReadHelper, for Configuration. If we inherit from Storage, then we can get Storage::Configuration and everything else needed from storage. You don't like the idea still? Just now we still have s3 mentions in StorageDeltaLake, for which you mentioned it is difficult to get rid of. So inheriting from needed storage will allow to get rid of s3 in that case?

Reply:
Currently, the class just has two templates: `Name` and `MetadataParser`. I think is OK to inherit from the Storage.

Reply:
But, if inherit from StorageS3, how can I initialize the base class, we need first parse the metadata and update configuration, so current use PImpl.

Reply:
hm, ok.
But let's just use Storage::Configuration instead of template Configuration, because definition of configuration will be visible even with uninitialised storage.

Reply:
> But, if inherit from StorageS3, how can I initialize the base class, we need first parse the metadata and update configuration, so current use PImpl.

This issue will be resolved if you use variadic argument for constructor of IStorageDataLake.

Reply:
I don't understand how to resolve if by variadic arguments. We still need to initialize StorageS3 in initialization list.

Reply:
something like:
```
template <class Storage, class ...Args>
explicit IStorageDataLake(Args && ...args) : Storage(std::forward<Args>(args)...) {}
```

Reply:
No, it's not about pass arguments, its becuase the arguments of IStorageDataLake different with StorageS3.

Reply:
Currently I see they are the same except for `getAdjustedConfiguration` being used for s3. But what is the problem with this then:
```
template <class Storage, class ...Args>
explicit IStorageDataLake(Storage::Configuration && configuration, Args && ...storage_args)
 : Storage(getAdjustedConfiguration(), std::forward<Args>(args)...) {}
```

Reply:
We need first `updateConfiguration` then `getAdjustedConfiguration`...

Reply:
This is manageable the same way too.